EMsimulation face and edge conductivities

dccowan opened the pull request:
#### Summary

This pull request covers the addition of face and edge conductivities to FDEM and TDEM simulations. Because most of the properties are part of a base EMsimulation class, work can be done to add this functionality to static problems in the future. Due to the amount of invertible and non-invertible properties, we will make new simulation classes to handle this type of problem, and leave the original EM simulation classes as is.

#### PR Checklist
* [x] If this is a work in progress PR, set as a Draft PR
* [x] Linted my code according to the [style guides](https://docs.simpeg.xyz/content/getting_started/contributing/code-style.html).
* [x] Added [tests](https://docs.simpeg.xyz/content/getting_started/practices.html#testing) to verify changes to the code.
* [x] Added necessary documentation to any new functions/classes following the
      expect [style](https://docs.simpeg.xyz/content/getting_started/practices.html#documentation).
* [ ] Marked as ready for review (if this is was a draft PR), and converted 
      to a Pull Request
* [ ] Tagged ``@simpeg/simpeg-developers`` when ready for review.

#### What does this implement/fix?

* Add the ability to define face and edge conductivities on for 2D and 3D FDEM and TDEM simulations.
* Add the Jvec and Jtvec methods to allow differentiation with respect to these properties
* Add utilities to project surfaces to tensor and tree meshes

dccowan committed:
Create face and edge property mass matrices. Add tests

dccowan committed:
Add conductance to simulation and added analytic 1d layered earth tests.

dccowan committed:
add derivative and adjoint tests

dccowan committed:
adjoint

dccowan committed:
Merge branch 'main' into EMsimulationFaceProperties

# Conflicts:
#	SimPEG/base/pde_simulation.py
#	tests/base/test_mass_matrices.py

dccowan committed:
formatting

dccowan committed:
Merge branch 'main' into EMsimulationFaceProperties

dccowan committed:
preliminary TEM tetrahedral

dccowan committed:
Merge branch 'main' into EMsimulationFaceProperties

dccowan committed:
Merge branch 'main' into EMsimulationFaceProperties

dccowan committed:
Add inject active faces and active edges mappings

dccowan committed:
First pass at all necessary pieces for FDEM and TDEM forward modeling with face and edge conductivities

dccowan committed:
Analytic conductance: TDEM passes, FDEM only for cyl mesh

dccowan committed:
Adjoint tests for face edge conductivity pass. Sigma only.

dccowan committed:
TDEM adjoint tests pass

dccowan committed:
add FDEM forward tests for face conductivity

dccowan committed:
Add deriv and adjoint tests for FDEM

dccowan committed:
black and flake8

dccowan committed:
fix style checks

dccowan committed:
add noqa exception

dccowan commented:
With regards to docstrings, this will be done when general docstrings for EM modules are created for the API.

dccowan committed:
add some docstrings

dccowan committed:
Merge branch 'main' into EMsimulationFaceProperties

dccowan committed:
formatting

dccowan committed:
one for formatting

dccowan committed:
one more. I know.....precommit

santisoler reviewed (changes requested):
Great job @dccowan in this PR! A ton of work done here, congrats!

I think this looks great. I left some comments below. Mostly around the following subjects:
- using numpydoc style in every docstring,
- small edits to use inplace operators (`+=`, `*=`, etc) for efficiency,
- not need to use unittest since we are using pytest now.

Related to the tests, some of the could be simplified using some pytest capabilities like [parametrizations](https://docs.pytest.org/en/6.2.x/parametrize.html#pytest-mark-parametrize-parametrizing-test-functions). I left one example of how I would refactor a test class. Feel free to apply similar changes to other tests. No need to do so to merge this PR though.

Also regarding tests, there are a few lines that aren't being tested (according to the coverage provided by CI). It would be nice to add particular test functions to test them.

Let me know what do you think!

dccowan committed:
Merge branch 'main' into EMsimulationFaceProperties

dccowan commented on SimPEG/electromagnetics/frequency_domain/fields.py:
```diff
@@ -622,6 +622,94 @@ def _charge_density(self, eSolution, source_list):
+        """
+        Current density from eSolution
+
+        :param numpy.ndarray eSolution: field we solved for
+        :param list source_list: list of sources
+        :rtype: numpy.ndarray
+        :return: current density
```
docstring inherited from parent class

dccowan commented on SimPEG/electromagnetics/frequency_domain/fields.py:
```diff
@@ -622,6 +622,94 @@ def _charge_density(self, eSolution, source_list):
+        This includes derivatives for volume, face and/or edge conductivities
+        depending on whether ``sigmaMap``, ``tauMap`` and/or ``kappaMap`` are set.
+
+        :param SimPEG.electromagnetics.frequency_domain.sources.BaseFDEMSrc src: source
+        :param numpy.ndarray v: vector to take product with
+        :param bool adjoint: adjoint?
+        :rtype: numpy.ndarray
+        :return: product of the current density derivative with respect to the
+            inversion model with a vector
```
docstring inherited from parent class

dccowan commented on SimPEG/electromagnetics/frequency_domain/fields.py:
```diff
@@ -952,6 +1040,138 @@ def _charge_density(self, bSolution, source_list):
+        \mathbf{M_e \, J} = \left ( \mathbf{M_{e\sigma} + M_{e\tau}
+        + M_{e\kappa}} \right ) \mathbf{e}
+
+    :param discretize.base.BaseMesh mesh: mesh
+    :param SimPEG.electromagnetics.frequency_domain.SurveyFDEM.Survey survey: survey
```
done

dccowan commented on SimPEG/electromagnetics/frequency_domain/fields.py:
```diff
@@ -952,6 +1040,138 @@ def _charge_density(self, bSolution, source_list):
+        """
+        Secondary electric field from bSolution
+
+        :param numpy.ndarray bSolution: field we solved for
+        :param list source_list: list of sources
+        :rtype: numpy.ndarray
+        :return: secondary electric field
```
inherited from parent

dccowan commented on SimPEG/electromagnetics/frequency_domain/fields.py:
```diff
@@ -952,6 +1040,138 @@ def _charge_density(self, bSolution, source_list):
+        e = self._edgeCurl.T * (self._MfMui * bSolution)
+        for i, src in enumerate(source_list):
+            s_e = src.s_e(self.simulation)
+            e[:, i] = e[:, i] + -s_e
```
done

dccowan commented on SimPEG/electromagnetics/frequency_domain/fields.py:
```diff
@@ -952,6 +1040,138 @@ def _charge_density(self, bSolution, source_list):
+        Derivative of the electric field with respect to the thing we solved
+        for
+
+        :param SimPEG.electromagnetics.frequency_domain.sources.BaseFDEMSrc src: source
+        :param numpy.ndarray v: vector to take product with
+        :param bool adjoint: adjoint?
+        :rtype: numpy.ndarray
+        :return: product of the derivative of the electric field with respect
+            to the field we solved for with a vector
```
inherited from parent

dccowan commented on SimPEG/electromagnetics/frequency_domain/fields.py:
```diff
@@ -952,6 +1040,138 @@ def _charge_density(self, bSolution, source_list):
+                    + self.__MeTau
+                    + self.__MeKappa
+                )
+                e[:, i] = MeyhatI * e[:, i]
```
done

dccowan commented on SimPEG/electromagnetics/frequency_domain/fields.py:
```diff
@@ -952,6 +1040,138 @@ def _charge_density(self, bSolution, source_list):
+
+            for i, src in enumerate(source_list):
+                s_e = src.s_e(self.simulation)
+                j[:, i] = j[:, i] - s_e
```
done

dccowan commented on SimPEG/electromagnetics/frequency_domain/fields.py:
```diff
@@ -952,6 +1040,138 @@ def _charge_density(self, bSolution, source_list):
+        """
+        Secondary current density from bSolution
+
+        :param numpy.ndarray bSolution: field we solved for
+        :param list source_list: list of sources
+        :rtype: numpy.ndarray
+        :return: primary current density
```
inherited from parent

dccowan committed:
some efficiency and docstrings

dccowan commented:
It looks like these were written this way for a reason

![image](https://github.com/simpeg/simpeg/assets/12970009/7aaa5da9-b2d5-4cc5-a927-d662d4211c47)

dccowan committed:
Revert "some efficiency and docstrings"

This reverts commit abaecfe927a6fb188444fb4b4c33958815e620f2.

dccowan committed:
Revert "Revert "some efficiency and docstrings""

This reverts commit 9d1d6411ab1cd2235e12df3f4b3d9cede3d4f215.

dccowan committed:
Allow operations of Zero() class

santisoler commented:
> It looks like these were written this way for a reason
> 
> ![image](https://private-user-images.githubusercontent.com/12970009/294319904-7aaa5da9-b2d5-4cc5-a927-d662d4211c47.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ0ODA3NzksIm5iZiI6MTcwNDQ4MDQ3OSwicGF0aCI6Ii8xMjk3MDAwOS8yOTQzMTk5MDQtN2FhYTVkYTktYjJkNS00Y2M1LWE5MjctZDY2MmQ0MjExYzQ3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTA1VDE4NDc1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWYzNWYxZTE0OGIyNTExMTQyNTcxMTFlZDdlYmExODYwNTcwZjVlOWVmNTg0ZGVmMzEwYmNmYzYyYTc0NTc2YTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.x6UDOI5CPnqyRxzWr-7Fmt-gbolH8kjf3z4qiieT9yU)

I see. Then nevermind those if they fail. I think the `Zero` object should support those operators, but that's out of the scope of this PR.

dccowan commented:
> > It looks like these were written this way for a reason
> > ![image](https://private-user-images.githubusercontent.com/12970009/294319904-7aaa5da9-b2d5-4cc5-a927-d662d4211c47.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ0ODA3NzksIm5iZiI6MTcwNDQ4MDQ3OSwicGF0aCI6Ii8xMjk3MDAwOS8yOTQzMTk5MDQtN2FhYTVkYTktYjJkNS00Y2M1LWE5MjctZDY2MmQ0MjExYzQ3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTA1VDE4NDc1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWYzNWYxZTE0OGIyNTExMTQyNTcxMTFlZDdlYmExODYwNTcwZjVlOWVmNTg0ZGVmMzEwYmNmYzYyYTc0NTc2YTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.x6UDOI5CPnqyRxzWr-7Fmt-gbolH8kjf3z4qiieT9yU)
> 
> I see. Then nevermind those if they fail. I think the `Zero` object should support those operators, but that's out of the scope of this PR.

Most of them should work. Just need to avoid ones that can be Zero

dccowan committed:
final details docs

dccowan commented on SimPEG/electromagnetics/frequency_domain/simulation.py:
```diff
@@ -426,6 +429,108 @@ def getRHSDeriv(self, freq, src, v, adjoint=False):
+        return (
+            self.getADeriv_sigma(freq, u, v, adjoint)
+            + self.getADeriv_mui(freq, u, v, adjoint)
+            # + self.getADeriv_permittivity(freq, u, v, adjoint)
```
This may be added in the future.

dccowan committed:
more concise

dccowan committed:
test comments round 1

dccowan committed:
Merge branch 'main' into EMsimulationFaceProperties

dccowan committed:
Merge branch 'main' into EMsimulationFaceProperties

dccowan committed:
some docstrings

jcapriot committed:
Merge branch 'main' into EMsimulationFaceProperties

jcapriot committed:
Merge branch 'main' into EMsimulationFaceProperties

dccowan committed:
Merge branch 'EMsimulationFaceProperties' of https://github.com/simpeg/simpeg into EMsimulationFaceProperties

dccowan committed:
Merge branch 'main' into EMsimulationFaceProperties

dccowan committed:
some docstring

jcapriot commented:
I've been running this myself (with a few modifications that I will suggest as a PR onto this one). The biggest few things should likely be:
* Settle on a name for the class. My preference would be to call it `Simulation3DHierarchicalFIELDNAME`, and refer to them as heirarchical elements, as this approach is the finite volume formulation of https://doi.org/10.1190/geo2017-0058.1, and there's no sense in coming up with new naming conventions.
* Mark the face element and edge element properties with `optional=True` so you are not required to supply either of them and can set them as `None` (and leave them initialized there).
    * Since they are optional you'll need to check if they are not None before accessing their mass matrices.
* Split the face properties and edge properties up into different base simulations.
    * You also don't need to re-write all of the code cleaning the matrices associated with conductivity/resistivity, just call `super().__setattr__(name, value)` as is done on the other classes (in general you could make a lot more use of `super()` calls throughout).

* The other thing is, I don't know if the relationship between current density and the electric fields are correct in this context? I'm not saying it is wrong I'm just saying I don't know if they're correct or not.

Another thing to consider it to possibly to just not have them be a separate class and mark them as experimental if they are set to something besides `None` (as is done with permittivity on the FDEM 3D electric field simulation.)

jcapriot committed:
Code simplifications and pytest updates

jcapriot committed:
testing updates

jcapriot committed:
update tdem analytic tests

jcapriot committed:
rerun pre-commit

jcapriot committed:
add mkvc back

jcapriot committed:
refactor SimPEG as simpeg

jcapriot committed:
add SimPEG stub

jcapriot committed:
more batch replacements

jcapriot committed:
Change to FutureWarning

jcapriot committed:
add stub test

jcapriot committed:
fix name to google cloud files

jcapriot committed:
next pass set `SimPEG` as project name, `simpeg` as the package name

jcapriot committed:
more updates for SimPEG -> simpeg

jcapriot committed:
update link targets in release notes

jcapriot committed:
updates in published examples

jcapriot committed:
directives

jcapriot committed:
More Project names

jcapriot committed:
more project names

jcapriot committed:
more Project names

jcapriot committed:
more project names...

jcapriot committed:
tutorial project names

jcapriot committed:
a bit more cleanup

jcapriot committed:
explicitly list the `SimPEG.py` as a py_module

jcapriot committed:
Merge branch 'main' into lowercase_simpeg

jcapriot committed:
Merge branch 'lowercase_simpeg' into em_hierarchical

dccowan committed:
Merge branch 'main' into EMsimulationFaceProperties

# Conflicts:
#	simpeg/electromagnetics/time_domain/simulation.py
#	tests/base/test_mass_matrices.py
#	tests/em/fdem/inverse/adjoint/test_FDEM_adjointEB.py
#	tests/em/fdem/inverse/derivs/test_FDEM_derivs.py

dccowan committed:
Merge branch 'main' into EMsimulationFaceProperties

# Conflicts:
#	SimPEG/electromagnetics/static/self_potential/simulation.py
#	SimPEG/electromagnetics/static/self_potential/sources.py
#	SimPEG/electromagnetics/static/spontaneous_potential/simulation.py
#	SimPEG/electromagnetics/static/spontaneous_potential/sources.py
#	simpeg/electromagnetics/static/spontaneous_potential/simulation.py
#	simpeg/electromagnetics/static/spontaneous_potential/sources.py

jcapriot committed:
temporary move

jcapriot committed:
unmove

jcapriot committed:
Merge branch 'main' into em_hierarchical

jcapriot committed:
Merge remote-tracking branch 'simpeg/EMsimulationFaceProperties' into em_hierarchical

dccowan committed:
Merge branch 'main' into EMsimulationFaceProperties

dccowan committed:
Merge branch 'main' into EMsimulationFaceProperties

dccowan committed:
Run black

jcapriot committed:
Suggested changes for Hierarchical EM implementation (#1411)

#### Summary
Implements my suggestions in #1296. 

#### PR Checklist
* [ ] If this is a work in progress PR, set as a Draft PR
* [ ] Linted my code according to the [style
guides](https://docs.simpeg.xyz/content/getting_started/contributing/code-style.html).
* [ ] Added
[tests](https://docs.simpeg.xyz/content/getting_started/practices.html#testing)
to verify changes to the code.
* [ ] Added necessary documentation to any new functions/classes
following the
expect
[style](https://docs.simpeg.xyz/content/getting_started/practices.html#documentation).
* [ ] Marked as ready for review (if this is was a draft PR), and
converted
      to a Pull Request
* [ ] Tagged ``@simpeg/simpeg-developers`` when ready for review.

#### What does this implement/fix?
* Rename to Hierarchical simulations
* Simplifies code for the hierarchical simulations (and fields) by
making use of inheritance.

#### Additional information
Also adapted some relevant tests to pytest parameterized formats.

dccowan committed:
Merge branch 'main' into EMsimulationFaceProperties

# Conflicts:
#	SimPEG/electromagnetics/frequency_domain/fields.py
#	SimPEG/electromagnetics/frequency_domain/simulation.py
#	SimPEG/electromagnetics/time_domain/simulation.py
#	SimPEG/fields.py
#	SimPEG/maps.py
#	tests/em/fdem/inverse/adjoint/test_FDEM_adjointEB.py
#	tests/em/fdem/inverse/adjoint/test_FDEM_adjointHJ.py
#	tests/em/fdem/inverse/derivs/test_FDEM_derivs.py
#	tests/em/tdem/test_TDEM_DerivAdjoint.py

dccowan committed:
Update SimPEG/electromagnetics/natural_source/utils/analytic_1d.py

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

dccowan committed:
black for autoreview suggestions

dccowan committed:
Merge branch 'EMsimulationFaceProperties' of https://github.com/simpeg/simpeg into EMsimulationFaceProperties

dccowan committed:
fix broken test utility

dccowan committed:
black for unrelated files

jcapriot committed:
Merge branch 'main' into em_hierarchical

jcapriot committed:
lowercase all in git history

dccowan committed:
Merge branch 'main' into EMsimulationFaceProperties

jcapriot committed:
Merge branch 'main' into em_hierarchical

jcapriot committed:
Merge branch 'em_hierarchical' into EMsimulationFaceProperties

jcapriot committed:
Revert "Merge branch 'em_hierarchical' into EMsimulationFaceProperties"

This reverts commit 293c237c6905a0c3fb14b76713e49b6659827abd, reversing
changes made to 81357b4fbe257c2fbc700ee445e4700bc8e3ddd2.

jcapriot committed:
fix case

jcapriot committed:
fix merge

jcapriot committed:
more merge fixes

jcapriot committed:
Merge branch 'main' into EMsimulationFaceProperties

dccowan committed:
Merge branch 'main' into EMsimulationFaceProperties

# Conflicts:
#	simpeg/__init__.py
#	simpeg/base/pde_simulation.py
#	simpeg/electromagnetics/frequency_domain/simulation.py
#	simpeg/electromagnetics/natural_source/simulation.py
#	simpeg/electromagnetics/static/spontaneous_potential/__init__.py
#	simpeg/electromagnetics/time_domain/sources.py
#	tests/base/test_base_pde_sim.py

dccowan committed:
update deprecated method name

dccowan committed:
Remove unused imports and fix typo in called method

dccowan committed:
Merge branch 'main' into EMsimulationFaceProperties

# Conflicts:
#	simpeg/electromagnetics/natural_source/simulation.py

dccowan committed:
Formatting. Yes, i know, pre-commit

dccowan committed:
Fix docstrings for base pde and mappings

dccowan committed:
Clean up analytic FDEM tests for face conductances
Add analytic TDEM tests for face conductances

dccowan committed:
Basic commit

dccowan committed:
Fix inheritance error

dccowan committed:
Ensure Adcinv is cleared for galvanic sources when model is updated

dccowan committed:
Merge branch 'main' into EMsimulationFaceProperties

dccowan committed:
Add derivative tests. Update deriv and adjoint test to use pytest

dccowan committed:
Modify DC hierarchical to linear regime so test passes

dccowan committed:
Add forward tests for hierarchical edge conductivities

dccowan committed:
Cleaning testing involving hierarchical physical properties for TDEM

dccowan committed:
docstrings for TDEM

dccowan committed:
FDEM simulation docstrings

dccowan committed:
docstrings for base EM simulation classes

dccowan committed:
Docstrings for 3D DCR resistivity simulation

dccowan committed:
update fdem and tdem simulation docstrings

dccowan committed:
Remove the addition of vector potential source for simplex meshes.

dccowan committed:
remove import

dccowan committed:
Clean up docstrings for mapping

dccowan committed:
updates to em tests

dccowan committed:
add test docstrings

dccowan committed:
docstrings for maps and base simulation

dccowan committed:
Merge branch 'main' into EMsimulationFaceProperties

dccowan committed:
Update to pytest. Cannot have repeated filename is subdirectory

dccowan committed:
base PDE tests. add random seed for repeatibility

dccowan commented:
Since this is a MASSIVE PR, we should break it into chunks. @jcapriot perhaps first, we can review the changes to the `base/pde_simulation.py`, `maps.py` and `regularization/regularization_mesh.py`. Once we are happy with those changes, we can then focus each simulation module one at a time. Agreed?

dccowan committed:
revert NSEM to main branch exactly

dccowan committed:
necessary to pass PR even though no actual changes to file.

lheagy commented on simpeg/base/pde_simulation.py:
```diff
@@ -668,10 +871,159 @@ def __setattr__(self, name, value):
         return toDelete
+
+
+@with_line_property_mass_matrices("kappa")
```
I made this comment in the meeting, but I would love for us to get away from naming physical properties by a greek letter. I realize that this is the standard now, but I would suggest that we reconsider the naming now, rather than anticipating a deprecation cycle for this. In that case I would suggest something like 
- `edge_integrated_conductance` or similar

lheagy commented on simpeg/base/pde_simulation.py:
```diff
@@ -668,10 +871,159 @@ def __setattr__(self, name, value):
+
+    kappa, kappaMap, kappaDeriv = props.Invertible(
+        "Electrical conductivity times thickness on face elements (S).",
+        optional=True,
```
shouldn't this be the edge integrated conductance? (Sm) ?

lheagy commented on simpeg/base/pde_simulation.py:
```diff
@@ -668,10 +871,159 @@ def __setattr__(self, name, value):
+
+@with_line_property_mass_matrices("kappa")
+class BaseElectricalEdgePropertyPDESimulation(BaseElectricalPDESimulation):
+    """Base class for simulations with edge electricl properties.
```
```suggestion
    """Base class for simulations with edge electrical properties.
```

lheagy commented on simpeg/base/pde_simulation.py:
```diff
@@ -668,10 +871,159 @@ def __setattr__(self, name, value):
+    tau, tauMap, tauDeriv = props.Invertible(
+        "Electrical conductivity times cross-sectional area on edge elements (Sm).",
+        optional=True,
+    )
```
I am a bit confused here - the class name is Face properties, but it looks like you are referring to an edge conductivity? (and again, lets get away from greek letters, especially the use of Tau, which is already overloaded in TDEM)

lheagy reviewed (commented):
A couple of quick comments for conversation. These are just based on what I briefly looked at in our meeting today

dccowan committed:
fix typos

dccowan committed:
Merge branch 'main' into EMsimulationFaceProperties

dccowan commented:
With the desire to move away from Greek letters for physical properties, I suppose we will have the following:
* sigma --> conductivity
* tau --> face_conductance
* kappa --> edge_integrated_conductivity

I think I would like to see how the property 'conductivity' is replaced, then I can update the hierarchical properties.

dccowan commented:
@lheagy @jcapriot I adopted property names 'face_conductance' and 'edge_area_conductance' to replace 'tau' and 'kappa'. However, the matrices that are constructed are still named _MeTau and _MeKappa, for example.

Basically I allowed the decorators in base_pde_simulation.py to accept both a `property_name`, which the and an optional `alias_name`. The property name (e.g. 'face_conductance') defines the name of the property that users interact with. The alias name is used to avoid generating matrices with names like _Me_face_conductance_I.

Let me know what you think.
